launch_handler.md

```diff
@@ -0,0 +1,283 @@
+  opened in a browser tab while already opened in an app window).
+- Configuring whether link navigations into the scope of a web app launch the
+  web app (this is out of scope and may be handled by a future version of the
+  [Declarative Link Capturing][dlc-explainer] spec).
+- Multi-document-instance web apps: a web app that opens documents in their own
+  instances but wishes to refocus an already open document instead of opening
+  duplicate instances for it. This would instead by handled by the [service
+  worker `launch` event][sw-launch-explainer].
+
+
+
+## Background
+
+- There are several ways for a web app window to be opened:
+  - [File handling](https://github.com/WICG/file-handling/blob/main/explainer.md)
+  - [In scope link capturing](https://github.com/WICG/sw-launch/blob/master/declarative_link_capturing.md)
+  - [Note taking](https://wicg.github.io/manifest-incubations/index.html#note_taking-member)
+  - Platform specific app launch surface
+  - [Protocol handling](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/URLProtocolHandler/explainer.md)
+  - [Share target](https://w3c.github.io/web-share-target/)
+  - [Shortcuts](https://www.w3.org/TR/appmanifest/#dfn-shortcuts)
+
+  Web apps launched via these triggers will open in a new or existing app window
+  depending on the user agent platform. There is currently no mechanism for the
+  web app to configure this behaviour.
+
+- This is a refactor of the [Declarative Link Capturing][dlc-explainer]
+  explainer with a reduced behaviour scope. This does not prescribe link
+  capturing as a means of launching a web app, instead it describes how the
+  launch of a web app (via link capturing or any other means) may be configured
+  by the manifest and service worker.
+
+
+## Proposal
+
+
+### `LaunchQueue` and `LaunchParams` interfaces
+
+Add two new interfaces; `LaunchParams` and `LaunchQueue`. Add a global instance
+of `LaunchQueue` to the `window` object named `launchQueue`.
+
+Whenever a web app is launched (via any launch trigger) a `LaunchParams` object
+will be enqueued in the `launchQueue` global `LaunchQueue` instance for the
+browsing context that handled the launch. Scripts can provide a single
+`LaunchConsumer` function to receive enqueued `LaunchParams`.
+
+This functions similarly to an event listener but avoids the problem where
+scripts may "miss" events if they're too slow to register their event listeners,
+this problem is particularly pronounced for launch events as they occur
+during the page's initialization. `LaunchParams` are buffered indefinitely until
+they are consumed.
+
+```
+[Exposed=Window] interface LaunchParams {
+  readonly attribute DOMString? targetURL;
+};
+
+callback LaunchConsumer = any (LaunchParams params);
+
+[Exposed=Window] interface LaunchQueue {
+  void setConsumer(LaunchConsumer consumer);
+};
+
+partial interface Window {
+  readonly attribute LaunchQueue launchQueue;
```

Comment:
I'd be cautious about adding things to the Window object. I don't have a better suggestion of where to put it, but doing this can be risky because it can clash with userland code.

Reply:
I think the main risk is if code does something like this:
```
launch_queue.js:
if (!window.launchQueue)
  window.launchQueue = new SomeLaunchQueueClass();


main.js:
window.launchQueue.someMethod(someParam);


index.html:
<script src="launch_queue.js"></script>
<script src="main.js"></script>
```
Do you know of any other code patterns that would break?